pages/api/submit.ts

```diff
@@ -19,44 +38,86 @@ async function serverSubmit(req: NextApiRequest, res: NextApiResponse) {
-  }
-  // A switch statement so later on we can do form value validation on specific
-  // fields based on the current page.
   switch (currentPage) {
     case "personal":
-      page = "location";
+      // Validate the form values for this specific step.
+      errors = validatePersonalFormData({}, newSubmittedValues);
+      if (isEmpty(errors)) {
+        page = "location";
+      }
       break;
     case "location":
-      page = "workAddress";
+      addresses = constructAddresses(newSubmittedValues);
+      errors = validateAddressFormData({}, addresses);
+      if (isEmpty(errors)) {
+        if (newSubmittedValues.location !== "nyc") {
+          page = "workAddress";
+        } else {
+          page = "account";
+        }
+      }
       break;
     case "workAddress":
-      page = "address-verification";
-      // TODO: Do address API call here.
-      break;
-    case "addressVerification":
-      page = "account";
+      // We need to add the existing home address values since
+      // `validateAddressFormData` checks for both home and work addresses.
+      addresses = constructAddresses({
+        ...newSubmittedValues,
+        ...existingValues,
+      });
+      errors = validateAddressFormData({}, addresses);
+      if (isEmpty(errors)) {
+        page = "account";
+      }
       break;
     case "account":
-      page = "review";
+      errors = validateAccountFormData({}, newSubmittedValues);
+      if (isEmpty(errors)) {
+        page = "review";
+      }
       break;
     case "review":
-      page = "congrats";
-      // TODO: Call the `createPatron` function with all the form values here.
+      try {
+        // All the data values are in the `existingValues` variable since
+        // no new data is submitted on the "review" page. When JS is not on,
+        // checkbox values are submitted as "on" and "off". We need to convert
+        // the values into booleans for the API to understand.
+        for (const [key, value] of Object.entries(existingValues)) {
+          if (value === "on") {
+            existingValues[key] = true;
+          } else if (value === "off") {
+            existingValues[key] = false;
+          }
+        }
+        // If the API call was successful and a patron account was created,
+        // go to the congrats page and display the results.
+        results = await callPatronAPI(existingValues);
+        page = "congrats";
+      } catch (error) {
+        errors = error;
+      }
       break;
     default:
       page = "new";
       break;
   }
 
-  return res.redirect(`/library-card/${page}?newCard=true${queryValues}`);
+  newSubmittedValues = { ...existingValues, ...newSubmittedValues };
+  queryValues =
+    createQueryParams(newSubmittedValues) +
+    createNestedQueryParams(errors, "errors") +
+    createNestedQueryParams(results, "results");
+
+  return res.redirect(
+    `/library-card/${page}?${encodeURI(`newCard=true${queryValues}`)}`
+  );
```

Comment:
Here all the form submission values, API results, and errors get converted to a key/value URL query string. The app pages read the query values, converts it to js objects for API results and errors, and puts those values in the app's store and state.